File: Stack.rb
# Represents a stack.
# Utilizes first in last out.
# All operations are constant time
class Stack
  attr_reader :size

  # No arg Constructor
  def initialize
    clear
  end

  # adds an item to the stack
  def push(element)
    @head = Node.new(element, @head)
    @size += 1
  end

  # Returns the item at the top
  def top
    raise ArgumentError, 'NoSuchElementException' if @head.nil?
    @head.element
  end

  # Returns the item at the top and removes it from the stack
  def pop
    raise ArgumentError, 'NoSuchElementException' if @head.nil?
    return_value = @head.element
    @head = @head.previous
    @size -= 1
    return_value
  end

  def clear
    @head = nil
    @size = 0
  end

  # Returns true if the stack is empty
  def empty?
    @size.zero?
  end

  private

  # Represents a single node in a stack
  class Node
    attr_reader :element, :previous
    def initialize(element, previous = nil)
      @element = element
      @previous = previous
    end
  end
end

File: Stack.cpp
                                                                                                                                                                                                                                                                                                                                                              #include<iostream>

using namespace std;

template <class E>
class Node{
  public:
    Node(E data);
    Node(E data, Node *prev);
    E getData();
  private:
    E element;
    Node<E>* previous;
};

template <class E>
Node<E>::Node(E data, Node<E>* prev){
  element = data;
  previous = prev;
}

template <class E>
E Node<E>::getData(){
  return element;
}

int main(){
  Node<int> node(10, 0);
  Node<int> node2(20, &node);
  cout << node.getData() << endl << node2.getData();
  return 0;
}
